lib/endpoints/vnc.js

```diff
@@ -0,0 +1,304 @@
+    this.next();
+
+    /* Now we continue on to use the websocket. */
+    S.gotoState('getport');
+};
+
+VNCConnectionFSM.prototype.state_getport = function state_getport(S) {
+    var vm = this.req.machine;
+    var uri = '/servers/' + vm.compute_node + '/vms/' + vm.id + '/vnc';
+    var self = this;
+    this.req.sdc.cnapi.get(uri, S.callback(function gotVNCDetails(err, obj) {
+        if (err) {
+            self.log.error(err, 'failed to fetch VM VNC details from CNAPI');
+            self.err = new restify.InternalServerError('Failed to retrieve ' +
+                'VNC socket details');
+            S.gotoState('error');
+            return;
+        }
+        if (typeof (obj.host) !== 'string' || typeof (obj.port) !== 'number') {
+            self.log.error({ obj: obj }, 'CNAPI returned invalid VM VNC obj');
+            self.err = new restify.InternalServerError('Failed to retrieve ' +
+                'VNC socket details');
+            S.gotoState('error');
+            return;
+        }
+        self.host = obj.host;
+        self.port = obj.port;
+        self.log = self.log.child({ vncHost: obj.host, vncPort: obj.port });
+        self.log.debug('cnapi returned address for vnc');
+        S.gotoState('connect');
+    }));
+    S.on(this.ws, 'error', function onWsError(err) {
+        S.gotoState('error');
+    });
+};
+
+VNCConnectionFSM.prototype.state_error = function state_error(S) {
+    this.log.warn(this.err, 'vnc connection exited with error');
+    if (this.ws) {
+        try {
+            this.ws.end(JSON.stringify({ type: 'error', error: this.err }));
+        } catch (ex) {
+            this.ws.destroy();
+        }
+    }
+    if (this.socket) {
+        this.socket.destroy();
+    }
+};
+
+VNCConnectionFSM.prototype.state_connect = function state_connect(S) {
+    var self = this;
+
+    S.on(this.ws, 'error', function connectWsError(err) {
+        self.err = err;
+        S.gotoState('error');
+    });
+    S.on(this.ws, 'end', function connectWsEnd() {
+        S.gotoState('error');
+    });
+
+    this.socket = net.createConnection({
+        allowHalfOpen: true,
+        host: this.host,
+        port: this.port
+    });
+
+    S.on(this.socket, 'connect', function connected() {
+        S.gotoState('connected');
+    });
+    S.on(this.socket, 'error', function connectSockErr(err) {
+        self.log.error(err, 'failed to connect to VNC endpoint');
+        self.err = new restify.InternalServerError('Failed to connect to ' +
+                'VNC server');
+        S.gotoState('error');
+    });
+    S.timeout(5000, function connectTimeout() {
+        self.log.error('timeout while connecting to VNC endpoint');
+        self.err = new restify.InternalServerError('Timeout while connecting ' +
+            'to VNC server');
+        S.gotoState('error');
+    });
+};
+
+VNCConnectionFSM.prototype.state_connected = function state_connected(S) {
+    var self = this;
+    this.socket.setNoDelay(true);
```

Comment:
Isn't `setNoDelay` already called during the ws.upgrade ?

Reply:
There's two sockets in play here: `client <-((web) socket)-> cloudapi <-(socket)-> CN` and no delay is desired for both. This is for the `cloudapi <-> CN` socket. The call in `state_upgrade` is the `client <-> cloudapi` websocket.